drivers: wifi: siwx91x: Add support for TWT

silabs-MuzaffarA opened the pull request:
- Implemented `siwx91x_set_twt_setup` and `siwx91x_set_twt_teardown` for handling TWT setup and teardown requests.
- Added validation for TWT parameters, including negotiation type, device state, and operation type.
- Integrated TWT functionality into the `wifi_mgmt_ops` structure for seamless management via the Zephyr Wi-Fi API.

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
+{
+	switch (z_req_cmd) {
+	case WIFI_TWT_SETUP_CMD_REQUEST:
+		return 0;
```
use macros. REQUEST_TWT, SUGGEST_TWT
For demand, use 2, as we need to get this fix in SDK.

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
+
+		if (params->teardown.teardown_all) {
+			twt_req.twt_flow_id = 0xFF;
+		}
```
twt_flow_id for individual teardown (not teardown_all)

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
+	sl_status_t status = SL_STATUS_OK;
+	sl_wifi_twt_request_t twt_req;
+	int ret;
+
```
check for interface disabled mode ?

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
+		if (params->teardown.teardown_all) {
+			twt_req.twt_flow_id = 0xFF;
+		}
+
```
check if connected with AP

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
 	return 0;
 }
 
+static int convert_z_sl_req_type(enum wifi_twt_setup_cmd z_req_cmd)
```
convert_z_sl_twt_req_type

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
+		if (params->setup.implicit) {
+			/* explicit twt is not supported */
+			twt_req.implicit_twt = 1;
+		}
```
as explicit TWT is not supported, return not supported.

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
+	twt_req.negotiation_type = 0;
+
+	if (params->operation == WIFI_TWT_SETUP) {
+		twt_req.twt_enable = 1;
```
if responder, return not supported

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -743,6 +743,114 @@ static int siwx91x_dev_init(const struct device *dev)
+
+	if (params->operation == WIFI_TWT_SETUP) {
+		twt_req.twt_enable = 1;
+		twt_req.wake_int_exp = params->setup.twt_exponent;
```
check the range for exponent

jhedberg commented:
@muahmed-silabs please prefix your PR and commit message titles so that it's clear what subsystem or driver the PR targets. E.g. "drivers: wifi: siwx91x: ...". I fixed this now for the PR title, but you need to fix it for the commit message.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.h:
```diff
@@ -14,6 +14,12 @@
 
+#define REQUEST_TWT 0
+#define SUGGEST_TWT 1
+#define DEMAND_TWT  2
```
I believe these values are exclusive, so define an enum instead.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -787,6 +787,120 @@ static int siwx91x_dev_init(const struct device *dev)
+static int siwx91x_set_twt(const struct device *dev, struct wifi_twt_params *params)
+{
+	sl_status_t status = SL_STATUS_OK;
+	sl_wifi_twt_request_t twt_req;
```
Fill this struct this 0 during declaration:
```
    sl_wifi_twt_request_t twt_req = { };
```
... so you won't have to patch your code if some new field are introduced.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -787,6 +787,120 @@ static int siwx91x_dev_init(const struct device *dev)
+		}
+		if (TWT_WAKE_INTERVAL_TO_256US(params->setup.twt_wake_interval) > UINT8_MAX) {
+			twt_req.wake_duration_unit = 1;
+			twt_req.wake_duration = TWT_WAKE_INTERVAL_TO_1024US(params->setup.twt_wake_interval);
```
`TWT_WAKE_INTERVAL_TO_1024US()` is probably useless. `params->setup.twt_wake_interval / 1024` seems fine.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -787,6 +787,120 @@ static int siwx91x_dev_init(const struct device *dev)
+		} else {
+			params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+			return -ENOTSUP;
+		}
```
I suggest to manage all the `-ENOTSUP` cases at the beginning of the function. Thus, the control flow for the nominal case will be easier to read.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -787,6 +787,120 @@ static int siwx91x_dev_init(const struct device *dev)
+	twt_req.twt_flow_id = params->flow_id;
+	twt_req.negotiation_type = 0;
+
+	if (params->operation == WIFI_TWT_SETUP) {
```
I feel `WIFI_TWT_SETUP` and `WIFI_TWT_TEARDOWN` are rather independent. Did you considered to introduce `siwx91x_set_twt_setup()` and  `siwx91x_set_twt_teardown()`? (I assume `twt_req.req_type` is only used with `WIFI_TWT_SETUP`?)

muahmed-silabs commented on drivers/wifi/siwx91x/siwx91x_wifi.h:
```diff
@@ -14,6 +14,12 @@
+#define SUGGEST_TWT 1
+#define DEMAND_TWT  2
+#define TWT_WAKE_INTERVAL_TO_256US(x)  ((x) >> 8)
+#define TWT_WAKE_INTERVAL_TO_1024US(x) ((x) >> 10)
```
I agree

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+
+	__ASSERT(params, "params cannot be a NULL");
+
+	sl_wifi_interface_t interface = sl_wifi_get_default_interface();
```
PLace this declaration with the others.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+	twt_req.req_type = ret;
+	/*  Device supports Individual TWT only */
+	twt_req.twt_flow_id = params->flow_id;
+	twt_req.negotiation_type = 0;
```
You have already assigned 0 to `negotiation_type`.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+	twt_req.twt_flow_id = params->flow_id;
+	twt_req.negotiation_type = 0;
+
+	if (params->operation == WIFI_TWT_SETUP) {
```
Now, you can simplify that:
```
    switch (params->operation) {
    case WIFI_TWT_SETUP:
        return siwx91x_set_twt_setup(params, &twt_req);
    case WIFI_TWT_TEARDOWN:
        return siwx91x_set_twt_teardown(params, &twt_req);
    default:
       return -ENOTSUP;
    }
```
(You can replace `switch ... case` by` if (...) { } else { }` if you prefer)

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+
+static int siwx91x_set_twt_setup(struct wifi_twt_params *params, sl_wifi_twt_request_t *twt_req)
+{
+	sl_status_t status = SL_STATUS_OK;
```
You don't have to initialize this variable since you are going to overwrite it anyway (rationale: the reader will try to look for where this value is used).

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -28,6 +28,12 @@ LOG_MODULE_REGISTER(siwx91x_wifi);
+enum siwx91x_wifi_twt_types {
+	REQUEST_TWT,
+	SUGGEST_TWT,
+	DEMAND_TWT,
```
This enum need binary compatibility with the wiseconnect API. So explicitly assign values to the symbols.

In fact, I even wonder if it make sense to define this enum. If WiseConnect does not provide it, just return raw values in `siwx91x_convert_z_sl_twt_req_type()`.

muahmed-silabs replied:
I understand, but I prefer using named variables (could be useful in searches). Will raise a request for WiseConnect to provide this enum in the future.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+
+static int siwx91x_set_twt(const struct device *dev, struct wifi_twt_params *params)
+{
+	sl_wifi_twt_request_t twt_req = { };
```
I feel we don't really need to declare this struct here. We could declare local instances in `siwx91x_set_twt_teardown()` and `siwx91x_set_twt_setup()` (rationale: it is more complex to follow the data flow when a struct is initialized from multiple place, especially if you consider `twt_flow_id` which modified on several places).

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+{
+	sl_status_t status = SL_STATUS_OK;
+
+	twt_req->twt_enable = 0;
```
You have already assigned 0 to `twt_enable` (but since this has been made in the caller function, this is not easy to follow).

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+		/* explicit twt is not supported */
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
```
This is an exception case. So:
```
    /* explicit twt is not supported */
    if (!params->setup.implicit) {
		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
		return -ENOTSUP;
    }
    twt_req->implicit_twt = 1;
```

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,144 @@ static int siwx91x_dev_init(const struct device *dev)
+	twt_req->un_announced_twt = !params->setup.announce;
+	twt_req->wake_int_mantissa = params->setup.twt_mantissa;
+	twt_req->wake_duration_unit = 0;
+	twt_req->wake_duration = params->setup.twt_wake_interval;
```
Once `twt_req` will be a local variable, you will be able to simplify these assignments:
```
    sl_wifi_twt_request_t twt_req = {
        .twt_enable = 1,
        .wake_int_exp = params->setup.twt_exponent,
        ....
    };
```

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,136 @@ static int siwx91x_dev_init(const struct device *dev)
+		.twt_flow_id = params->flow_id,
+		.twt_enable = 1,
+		.req_type = twt_req_type,
+	};
```
Avoid declarations after statements. You can relocate this init at the top of the function.

jhedberg replied:
IMO only the declaration could be moved, since there's a bunch of parameter value checking that happens after that, so assigning those values to the `twt_req` object *before* the checks seems a bit unintuitive.

jerome-pouiller replied:
This was the original code.

Personally, I think it is useless to assign 0 to the struct while the correct value is already known. However, this is matter of taste.

jhedberg replied:
Just declaring doesn't imply having to assign a value. Anyway, sorry for the drive-by comment - I wasn't aware of the history of the code & review.

muahmed-silabs replied:
I had placed it below the checks because I did not want the struct to be populated in case of any invalid values and early exit. I have moved this above for now, but let me know what you think.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,136 @@ static int siwx91x_dev_init(const struct device *dev)
+	} else if (params->operation == WIFI_TWT_TEARDOWN) {
+		return siwx91x_set_twt_teardown(params);
+	}
+	return 0;
```
If `params->operation` is not supported, we return 0. Is it what we want?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -28,6 +28,12 @@ LOG_MODULE_REGISTER(siwx91x_wifi);
 
 NET_BUF_POOL_FIXED_DEFINE(siwx91x_tx_pool, 1, _NET_ETH_MAX_FRAME_SIZE, 0, NULL);
 
+enum siwx91x_wifi_twt_types {
```
You don't need to give a name to this enum. You can keep it anonymous.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -967,6 +973,136 @@ static int siwx91x_dev_init(const struct device *dev)
+static int siwx91x_set_twt_teardown(struct wifi_twt_params *params)
+{
+	sl_status_t status = SL_STATUS_OK;
+	sl_wifi_twt_request_t twt_req;
```
You don't need to initialize status, but you need to initialize `twt_req` (otherwise you are going to pass random values to `sl_wifi_disable_target_wake_time()`):
```
sl_wifi_twt_request_t twt_req = { };
```

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -1067,6 +1073,139 @@ static int siwx91x_dev_init(const struct device *dev)
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+	return 0;
```
Compliance check complains. It expects to see:
```
	...
	}
	if (params->operation == WIFI_TWT_TEARDOWN) {
		return siwx91x_set_twt_teardown(params);
	}
	params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
	return -ENOTSUP;
```

muahmed-silabs replied:
Thank you. Updated accordingly.

silabs-MuzaffarA committed:
drivers: wifi: siwx91x: Add support for TWT

- Implemented `siwx91x_set_twt_setup` and `siwx91x_set_twt_teardown` for
  handling TWT setup and teardown requests.
- Added validation for TWT parameters, including negotiation type,
  device state, and operation type.
- Integrated TWT functionality into the `wifi_mgmt_ops` structure for
  seamless management via the Zephyr Wi-Fi API.

Signed-off-by: Muzaffar Ahmed <muzaffar.ahmed@silabs.com>

jhedberg commented:
@muahmed-silabs for the future, please always use `--keep-base` when doing `git rebase`, unless you need to fix merge conflicts or similar. This way the "Compare" link that GitHub provides will show a meaningful diff to the previous version of the PR.

jhedberg commented:
@muahmed-silabs please check the twister failures

muahmed-silabs commented:
@jhedberg I did check, it looks like the Twister issues are CI related.

jhedberg commented:
> @jhedberg I did check, it looks like the Twister issues are CI related.

I agree. Seems like breakage on the `main` branch, so we'll need to retrigger these tests once that has been fixed.

jerome-pouiller commented:
I believe the breakage has been fixed by a3f658454ab. I believe we need to rebase the PR.
